proteus/Gauges.py

```diff
@@ -267,18 +351,139 @@ def buildGaugeOperators(self):
+                l_i.difference_update(remove_list)
+                log("Proc %d has segments %s" % (proc, l_i), 9)
+                log("Global segments %s" % (global_segments), 9)
+            err = np.linalg.norm(original_v - segment_v)/np.linalg.norm(original_v)
+            if err > 1e-8:
+                msg = "Segmented line %s is different from original length by factor %e" % (str(endpoints), err)
+                log(msg, 3)
+                if err > 1e-4:
+                    raise FloatingPointError(msg)
+
+        segments = comm.scatter(local_segments)
+        return segments
+
+    def getMeshIntersections(self, line):
+        field, endpoints = line
+        # get Proteus mesh index for this field
+        field_id = self.fieldNames.index(field)
+        femFun = self.u[field_id]
+        mesh = femFun.femSpace.mesh
+        referenceElement = femFun.femSpace.elementMaps.referenceElement
+        if referenceElement.dim == 2 and referenceElement.nNodes == 3:
+            toPolyhedron = Geom.triangleVerticesToNormals
+        elif referenceElement.dim == 3 and referenceElement.nNodes == 4:
+            toPolyhedron = Geom.tetrahedronVerticesToNormals
+        else:
+            raise NotImplementedError("Unable to compute mesh intersections for this element type")
+        segments = Geom.getMeshIntersections(mesh, toPolyhedron, endpoints)
+        segments = self.pruneDuplicateSegments(endpoints, segments)
+
+        return segments
+
+
+    def buildLineGaugeOperators(self, lines, linesSegments):
+        """ Build the linear algebra operators needed to compute the line gauges.
+
+        The operators are collective since the line gauge measurements may be across multiple processors.
+
+        Unlike point gauges, each line has its own communicator and field associated with it
+        """
+
+        #create lineGaugesVec to store contributions to all lines from this process
+        self.lineGaugesVec = PETSc.Vec().create(comm=PETSc.COMM_SELF)
+        self.lineGaugesVec.setSizes(len(lines))
+        self.lineGaugesVec.setUp()
+
+        # create lineGaugeMats to store coefficients mapping contributions from each field
+        # to the line gauges
+        self.lineGaugeMats = []
+
+        if not self.isLineGauge:
+            return
+
+        # size of lineGaugeMats depends on number of local points for each field
+        for pointGaugesVec in self.pointGaugeVecs:
+            m = PETSc.Mat().create(comm=PETSc.COMM_SELF)
+            m.setSizes([len(lines), pointGaugesVec.getSize()])
+            m.setType('aij')
+            m.setUp()
+            self.lineGaugeMats.append(m)
+
+        # Assemble contributions from each point in each line segment
+
+        for lineIndex, (line, segments) in enumerate(zip(self.lines, linesSegments)):
+            field, endpoints = line
+            fieldIndex = self.fields.index(field)
+
+            # Trapezoid Rule to calculate coefficients here
```

Comment:
For general integrated line gauges I think we'll have to use quadrature based on the order of the finite element space.  We're effectively creating a 1D mesh and integrating the finite element function over it, so most of the functionality is already implemented in Quadrature and FemTools. We just haven't ever applied to this case. It's not a high priority.